nbqa/save_source.py

```diff
@@ -87,25 +89,87 @@ def _handle_magic_indentation(
+    # `SyntaxError`. Since we are interested in only checking the indentation
+    # of the current line magic, previous line magics should be replaced with
+    # some valid python token.
+    # `%time print("hello")` will become `magic; print("hello")`
+    cell_source = re.sub(r"%\w+", "magic;", cell_source)
+
+    if leading_space and not _is_src_code_indentation_valid(cell_source):
         # Remove the line magic indentation assuming these leading spaces
         # lead the source to have invalid indentation
         # Currently we don't check if the original source
         # code itself had IndentationError
-        replacement = magic_replacement.replacement_line
+        replacement = magic_replacement
 
-    if line_magic.endswith("\n"):
-        replacement += "\n"
+    if line_magic.endswith(NEWLINE):
+        replacement += NEWLINE
 
     return replacement
 
 
+def _extract_ipython_magic(magic: str, cell_source: Iterator[Tuple[int, str]]) -> str:
+    r"""Extract the ipython magic from the notebook cell source.
+
+    To extract ipython magic, we use `nbconvert.get_lines` because it can extract
+    the ipython magic statement that can span multiple lines.
+
+    .. code:: python
+
+        # example of ipython magic spanning multiple lines
+        %time result_pymc3 = eval_lda(\
+        transform_pymc3, beta_pymc3, docs_te.toarray(), np.arange(100)\
+        )
+
+    `nbconvert.filters.get_lines` has the capability to parse such line magics and
+    return the result as a single line with trailing backslash removed. `get_lines`
+    would return `%time result_pymc3 = eval_lda(        transform_pymc3, beta_pymc3,
+     docs_te.toarray(), np.arange(100)    )`
+
+    If the magic was spanning multiple lines, then we need to remove all trailing
+    backslashes introduced by previous runs of nbqa. Also we do need to preserve the
+    newline characters, otherwise tools like black will format the code again. Linters
+    like flake8, pylint will complain about line length. After we extract the code, we
+    should have string like below
+
+    .. code:: python
+
+        %time result_pymc3 =eval_lda(
+            transform_pymc3, beta_pymc3, docs_te.toarray(), np.arange(100)
+        )
+
+    Parameters
+    ----------
+    magic : str
+        First line of the ipython magic
+
+    cell_source: Iterator[Tuple[int, str]]
+        Iterator to the notebook cell source
+
+    Returns
+    -------
+    str
+        IPython line magic statement
+    """
+    if re.match(r"\s*%\w+", magic) is not None:
+        # Here we look for line magics spanning across multiple lines.
+        while magic.endswith(f"\\{NEWLINE}"):
+            try:
+                magic += next(cell_source)[1]
+            except StopIteration:  # pragma: nocover
+                # This scenario is a syntax error where a line magic
+                # ends with a backslash and does not have a next line.
+                break
```

Comment:
Sorry for the (too) late reply.
What I had in mind was something in the direction.
warnings.warn(
    """Unable to extract magic from cell x.y, please check that the cell produces the desired output.

If it does, please file an issue at nbqa, with an example to reproduce the problem.
""")

That way we could inform users that something went wrong w/o breaking the CI if the code is fine and we have an internal error.

Reply:
it's OK, this could go in 0.4.1, thanks for your input!